### PR TITLE
ci: Update release workflow

### PR DESCRIPTION
**Pull request by helayoty**
**Reason for Change**:
- Automate the release workflow.
- Use [Reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows#reusable-workflows-and-starter-workflows) instead of workflow_dispatch. This will enhance the usability of e2e pipeline.
- Remove unnecessary jobs.
- Fix goreleaser to pick the current tag.
- Add option to update k8s version for upcoming pipelines.

**Requirements**
- [ ] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

**Notes for Reviewers**:

### CODE REVIEW

File: .github/workflows/publish-acr-image.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  create-release:
+    runs-on: ubuntu-latest
+    environment: publish-mcr
+    needs: [ run-e2e-mcr ]
+    steps:
+      - name: 'Dispatch release tag'
+        uses: peter-evans/repository-dispatch@v3
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          event-type: create-release
```
Comment (ishaansehgal99) [inline .github/workflows/publish-acr-image.yml:79]: create-release event here triggers both create-release.yml (goreleaser) and helm-chart.yml (uploads chart to github page)

Is it intended to be triggered for both publish-acr image anfd publish-gh image?

File: .github/workflows/helm-chart.yml [hunk 2]
```diff
@@ -15,15 +13,15 @@ permissions:
...
     steps:
       - name: Checkout
-        uses: actions/checkout@v4
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
         with:
           submodules: true
           fetch-depth: 0
+          ref: ${{ github.event.client_payload.tag }}
 
       - name: Publish Workspace Helm chart
```
Comment (ishaansehgal99) [inline .github/workflows/helm-chart.yml:26]: Uploads helm chart to GH pages right? I can't see our charts here yet: https://azure.github.io/kaito/

File: .github/workflows/publish-gh-image.yml [hunk 3]
```diff
@@ -16,51 +14,45 @@ env:
...
-    env:
-      REGISTRY: ${{ needs.export-registry.outputs.registry }}
-    runs-on: ubuntu-20.04
-    environment: preset-env
-    steps:
       - id: get-tag
         name: Get tag
-        run: echo "IMG_TAG=$(echo ${{ github.event.pull_request.head.ref }} | tr -d release-)" >> $GITHUB_ENV
-      - uses: actions/checkout@v4
+        run: echo "release_tag=$(echo ${{ github.event.pull_request.head.ref }} | tr -d release-)" >> $GITHUB_OUTPUT
```
Comment (ishaansehgal99) [inline .github/workflows/publish-gh-image.yml:31]: ok this refers to the branch reference of the head of the pull request - got it

File: .github/workflows/publish-gh-image.yml [hunk 4]
```diff
@@ -16,51 +14,45 @@ env:
...
+
+      - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
         with:
           submodules: true
           fetch-depth: 0
-          ref: ${{ env.IMG_TAG }}
+          ref: ${{ steps.get-tag.outputs.release_tag }}
```
Comment (ishaansehgal99) [inline .github/workflows/publish-gh-image.yml:37]: So I get that this checks out this release_branch, if this modified branch doesn't exist yet this checkout will fail right?

File: .github/workflows/publish-acr-image.yml [hunk 5]
```diff
@@ -0,0 +1,80 @@
+name: Push image to ACR
+on:
+  repository_dispatch:
+    types: [ publish-mcr-image ]
```
Comment (ishaansehgal99) [inline .github/workflows/publish-acr-image.yml:4]: I am probably overlooking but where is the event being fired?

File: .github/workflows/publish-gh-image.yml [hunk 6]
```diff
@@ -1,10 +1,8 @@
 name: Create, Scan and Publish KAITO image
 on:
-  pull_request:
-    branches:
-      - main
-      - release-**
-    types: [ closed ]
+  workflow_run:
+    workflows: [ "Push image to ACR" ]
```
Comment (ishaansehgal99) [inline .github/workflows/publish-gh-image.yml:4]: Were we going to switch order such that GH workflow ran first?

File: .github/workflows/publish-gh-image.yml [hunk 7]
```diff
@@ -76,9 +68,30 @@ jobs:
...
+    needs: [run-e2e-gh-image]
+    steps:
       - name: 'Dispatch release tag'
         uses: peter-evans/repository-dispatch@v3
         with:
           token: ${{ secrets.GITHUB_TOKEN }}
-          event-type: release-tag
-          client-payload: '{"isRelease": true,"registry": "$${{ env.REGISTRY }}","tag": "v${{ env.IMG_TAG }}"}'
+          event-type: create-release
+          client-payload: '{"tag": "v${{ env.IMG_TAG }}"}'
```
Comment (ishaansehgal99) [inline .github/workflows/publish-gh-image.yml:97]: uses branch reference of head of pull request - got it

File: .github/workflows/publish-acr-image.yml [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    runs-on: ubuntu-latest
+    environment: publish-mcr
+    needs: [ run-e2e-mcr ]
+    steps:
+      - name: 'Dispatch release tag'
+        uses: peter-evans/repository-dispatch@v3
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          event-type: create-release
+          client-payload: '{"tag": "${{ github.event.client_payload.tag }}"}'
```
Comment (ishaansehgal99) [inline .github/workflows/publish-acr-image.yml:80]: Similiar to GH workflow, can this also use branch reference of head of PR for consistency...or no

File: .github/workflows/publish-acr-image.yml [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+permissions:
+  contents: write
+  packages: write
+
+env:
+  GO_VERSION: '1.20'
+  IMAGE_NAME: 'workspace'
+
+jobs:
+  build-publish-mcr-image:
```
Comment (Fei-Guo) [inline .github/workflows/publish-acr-image.yml:15]: A little confuse about the name. The file is publish-acr-image, and why mcr is here? or maybe your should change the file name.

File: .github/workflows/publish-acr-image.yml (hunk 5 above)
Comment (Fei-Guo) [inline .github/workflows/publish-acr-image.yml:4]: Why is it called publish-mcr-image? not publish-acr-image?

File: .github/workflows/publish-gh-image.yml [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+        if:  steps.check-tag.outputs.create_tag == 'true'
+        uses: actions/github-script@v7
+        with:
+          script: |
+            github.rest.git.createRef({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              ref: 'refs/tags/${{ steps.get-tag.outputs.tag }}',
+              sha: context.sha
+            })
```
Comment (ishaansehgal99) [inline .github/workflows/publish-gh-image.yml:74]: creates branch with tag right

File: .github/workflows/publish-gh-image.yml [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    runs-on: ubuntu-latest
+    environment: preset-env
+    needs: [ check-tag, run-e2e-gh-image ]
+    steps:
+      - name: 'Dispatch release tag'
+        uses: peter-evans/repository-dispatch@v3
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          event-type: publish-mcr-image
+          client-payload: '{"tag": "${{ needs.check-tag.outputs.tag }}"}'
```
Comment (ishaansehgal99) [inline .github/workflows/publish-gh-image.yml:151]: I think this should be needs.get-tag

File: .github/workflows/publish-gh-image.yml [hunk 12]
```diff
@@ -0,0 +1,151 @@
+name: Create, Scan and Publish KAITO image
+on:
+  workflow_dispatch:
+    inputs:
+      release_version:
+        description: 'tag to be created for this image (i.e. vxx.xx.xx)'
+        required: true
+  pull_request:
+    types: [ closed ]
```
Comment (ishaansehgal99) [inline .github/workflows/publish-gh-image.yml:9]: Every closed pull request triggers this pipeline?

**Review by ishaansehgal99 — APPROVED** [review]

File: .github/workflows/publish-gh-image.yml (hunk 11 above)
Comment (ishaansehgal99) [reply .github/workflows/publish-gh-image.yml:151]: my bad check-tag is job name and step name - this works

File: .github/workflows/publish-gh-image.yml (hunk 4 above)
Comment (ishaansehgal99) [reply .github/workflows/publish-gh-image.yml]: Gotchya I mixed up tag and branch

**Review by ishaansehgal99 — APPROVED** [review]

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**